server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -46,42 +75,112 @@ public void beforeRefresh() throws IOException {
+                        }
+                        try (GatedCloseable<SegmentInfos> segmentInfosGatedCloseable = indexShard.getSegmentInfosSnapshot()) {
+                            SegmentInfos segmentInfos = segmentInfosGatedCloseable.get();
+                            Collection<String> refreshedLocalFiles = segmentInfos.files(true);
 
-            Set<String> remoteFilesToBeDeleted = new HashSet<>();
-            // ToDo: Instead of deleting files in sync, mark them and delete in async/periodic flow (GitHub #3142)
-            filesUploadedToRemoteStore.stream().filter(file -> !localFiles.contains(file)).forEach(file -> {
-                try {
-                    remoteDirectory.deleteFile(file);
-                    remoteFilesToBeDeleted.add(file);
-                } catch (IOException e) {
-                    // ToDO: Handle transient and permanent un-availability of the remote store (GitHub #3397)
-                    logger.warn(() -> new ParameterizedMessage("Exception while deleting file {} from the remote segment store", file), e);
+                            List<String> segmentInfosFiles = refreshedLocalFiles.stream()
+                                .filter(file -> file.startsWith(IndexFileNames.SEGMENTS))
+                                .collect(Collectors.toList());
+                            Optional<String> latestSegmentInfos = segmentInfosFiles.stream()
+                                .max(Comparator.comparingLong(IndexFileNames::parseGeneration));
+
+                            if (latestSegmentInfos.isPresent()) {
+                                refreshedLocalFiles.addAll(SegmentInfos.readCommit(storeDirectory, latestSegmentInfos.get()).files(true));
+                                segmentInfosFiles.stream()
+                                    .filter(file -> !file.equals(latestSegmentInfos.get()))
+                                    .forEach(refreshedLocalFiles::remove);
+
+                                boolean uploadStatus = uploadNewSegments(refreshedLocalFiles);
+                                if (uploadStatus) {
+                                    remoteDirectory.uploadMetadata(
+                                        refreshedLocalFiles,
+                                        storeDirectory,
+                                        indexShard.getOperationPrimaryTerm(),
+                                        segmentInfos.getGeneration()
+                                    );
+                                    localSegmentChecksumMap.keySet()
+                                        .stream()
+                                        .filter(file -> !refreshedLocalFiles.contains(file))
+                                        .collect(Collectors.toSet())
+                                        .forEach(localSegmentChecksumMap::remove);
+                                }
+                            }
+                        } catch (EngineException e) {
```

Comment:
Since we swallow all exceptions, what happens if shard was promoted to primary but segments were not refreshed due to some failures? Should we go back to replica mode?

Reply:
With the current implementation, we retry segment upload in the next refresh.

If the upload fails, we don't want to demote primary as the failure could be due to issues with remote segment store itself.